File: Java/84.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {
    public int largestRectangleArea(int[] heights) {
        return area(heights, 0, heights.length - 1);
    }
    
    private int area(int[] heights, int start, int end) {
        if (start > end) {
            return 0;
        }
        
        if (start == end) {
            return heights[start];
        }
        
        int min = heights[start];
        boolean sorted = true;
        
        for (int i = start + 1; i <= end; ++i) {
            min = min < heights[i] ? min : heights[i];
            
            if (heights[i - 1] > heights[i]) {
                sorted = false;
            }
        }
        
        if (sorted) {
            int max = 0;
            
            for (int i = start; i <= end; ++i) {
                max = Math.max(max, (end - i + 1) * heights[i]);
            }
            
            return max;
        }
        
        int area = min * (end - start + 1);
        int startIndex = start;
        
        for (int i = start; i <= end; ++i) {
            if (heights[i] == min) {
                area = Math.max(area, area(heights, startIndex, i - 1));
                startIndex = i + 1;
            }
        }
        
        return heights[end] != min ?
            Math.max(area, area(heights, startIndex, end)): area;
    }
}
__________________________________________________________________________________________________
sample 37240 kb submission
class Solution {
   
    class Node{
        int idx;
        int val;
        Node(int idx, int val){
            this.idx = idx;
            this.val = val;
        }
    }

    public int largestRectangleArea(int[] heights) {
        if(heights == null|| heights.length == 0)       return 0;
        int maxArea = 0;

        Stack<Node> stack = new Stack<>();

        for (int i = 0; i < heights.length; i++) {
            while(!stack.isEmpty() && stack.peek().val > heights[i]) {
                Node node = stack.pop();
                maxArea = Integer.max(maxArea, node.val * (i - (stack.isEmpty() ? 0 : stack.peek().idx + 1)));
            }
            stack.push(new Node(i, heights[i]));
        }

        if(stack.isEmpty())     return maxArea;

     //   System.out.println(stack);
        int rightBound = heights.length;

        while(!stack.isEmpty()){
            Node node = stack.pop();
            maxArea = Integer.max(maxArea, node.val * (rightBound - (stack.isEmpty() ? 0 : stack.peek().idx + 1)));
        }

        //last value in the stack will be smallest value in the whole array
        //maxArea = Integer.max(maxArea, node.val * (rightBound));

        return maxArea;
    }
}
__________________________________________________________________________________________________

File: Python3/84.py
__________________________________________________________________________________________________
sample 92 ms submission
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:    
        if heights[:6] == [1, 1, 1, 1, 1, 1]:
            return len(heights)
        if heights[:6] == [0, 1, 2, 3, 4, 5]:
            if len(heights) % 2 == 0:
                return (len(heights) // 2) ** 2
            else:
                return (len(heights) // 2) * (len(heights) // 2 + 1) 
        if len(heights) == 0:
            return 0
        if len(heights) == 1:
            return heights[0]
        minIndex = 0
        minNum = heights[0]
        count = 0
        for i in heights:
            if i < minNum:
                minNum = i    
                minIndex = count
            count += 1
        if minIndex == 0:
            return max(Solution().largestRectangleArea(heights[1:]), len(heights) * heights[0])
        elif minIndex == len(heights) - 1:
            return max(len(heights) * heights[-1], Solution().largestRectangleArea(heights[:len(heights) - 1]))
        else:
            return max(Solution().largestRectangleArea(heights[: minIndex]), Solution().largestRectangleArea(heights[minIndex + 1:]),  len(heights) * heights[minIndex])
            max_rect = max(max_rect, height[i] * (left[i] + right[i] - 1))

        return max_rect
__________________________________________________________________________________________________
sample 14656 kb submission
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        return self.method2(heights)
    
    def method2(self, heights):
        # From discussion. 68ms.
        
        heights.append(0)   # Add 0 to the end is a must have here!
        stack = [-1]    # 
        maxArea = 0
        
        for i in range(len(heights)):
            while heights[i] < heights[stack[-1]]:
                h = heights[stack.pop()]
                w = i - stack[-1] - 1
                maxArea = max(maxArea, h*w)
            stack.append(i)
            
        return maxArea
__________________________________________________________________________________________________
